Develop

franz-delatorre opened the pull request:
Changes made are based on the initial review.
Changes made:
1. removed unused variables and imports
2. GameMap
* change `setGameMap()` to `setOpenRooms()`
* refactored `canMoveToAdjacentRoom()`
* changed `showMap()` for-loop to foreach

3. Inventory - removed `throws IndexOutOfBoundsException` for `getItem()`
4. UnitEquipment - changed equipment variable from HashMap to EnumMap
5. Skill package - removed skill list package and migrated all classes inside the said directory to skill package
6. Stats - removed argument for the EnumMap for the variable stat.
7. Narrative 
* renamed and refactored `getStrings()` to `getNarrative()`
8. BattleManager
* Revised javadoc for `toBattle()`
* refactored `setCooldown()`
9. Removed `isAlive` variable from Unit and added an `isAlive` variable in Health
* Reason: I think health should be the one to know that it is still alive 

# Architecture Revision
- Game Engine caters cartridges.
- Each cartridge will have different game mechanism. This will uniquely give each game its own signature. Example of which is the way the game is played, how it ends and etc. 
- Cartridges should have the same Game Type with the Game Engine.

Commit:
SUD: revised GameMap

Commit:
SUD: revised Room

Commit:
SUD: revised Game Narrative

Commit:
SUD: Refactored codes based on the recent review

Commit:
SUD: Version 2

Commit:
SUD: renamed heal variable

zmenriquez commented on src/game/engine/GameManager.java:
```diff
@@ -258,47 +254,53 @@ private void showCharacter() {
+                    }
                 }
             }
+        } catch (NullPointerException e) {
```
When will this be thrown?

zmenriquez commented on src/game/engine/GameManager.java:
```diff
@@ -258,47 +254,53 @@ private void showCharacter() {
+
+        //Checks if the item is already in the hero's inventory
+        if (inventoryService.heroInventoryContains(item)) return;
+        inventoryService.addItemToHeroInventory(item);
```
Why is that line included in `checkRoomItem`?  It doesn't fit the method name.

zmenriquez commented on src/service/BattleService.java:
```diff
@@ -118,6 +124,7 @@ private void useSkill(SkilledUnit user, Unit victim) {
         Skill skill = user.getSkill();
         System.out.println(ANSI_CYAN + user.getName() + " used " + skill.getName() + ANSI_BLACK);
         SkillHelper.useSkill(user, victim);
+        SkillHelper.useSkill(user, victim);
```
Lines 126 and 127 are the same. Is that intentional?

zmenriquez commented on src/service/BattleService.java:
```diff
@@ -161,17 +165,14 @@ private boolean fightersStillAlive() {
-            return 1;
-        }
+        if (hero.isAlive()) return 1;
         return 0;
```
You can use a ternary operator for this

Commit:
SUD: refactored GameManager and BattleService

franz-delatorre commented:
Have addressed the issues.

Commit:
SUD: refactored Narrative

Commit:
SUD: refactored Narrative and Game Narrative

zmenriquez commented on src/components/skill/HealSkill.java:
```diff
@@ -26,17 +25,16 @@ public void setHeal(int heal) {
+        int userMaxHealth = user.getHealth().getMaxHealth();
         int userCurrentHealth = user.getHealth().getCurrentHealth();
-        Health userhealth     = user.getHealth();
+        Health userHealth = user.getHealth();
```
```java
Health userHealth = user.getHealth();
int userMaxHealth = userHealth.getMaxHealth();
int userCurrentHealth = userHealth.getCurrentHealth();
```
Create `userHealth` first so you can use that to get the max and current health.

zmenriquez commented on src/dialogue/GameNarrative.java:
```diff
@@ -7,14 +7,21 @@
     public Narrative getNarrative(Room room) {
-        return gameNarrative.get(room);
+        if (gameNarrative.containsKey(room)) return gameNarrative.get(room);
+        return new GameNarrative(room).getNarrative(room);
```
Why create a new `GameNarrative`? What happens to the old one? Explain lang, wala ra ko kasabot haha

franz-delatorre replied:
Since some rooms don't have a Narrative, it will return null. So my solution for the null return is creating a new instance of a GameNarrative with the room not listed in the current GameNarrative. The new GameNarrative then passes an empty Narrative for that room search. I think using getOrDefault would also be good here.

Commit:
SUD: changed the directory and modified GameNarrative and HealSkill

franz-delatorre commented:
Have addressed the issue and also changed the directory.

zmenriquez commented on src/com/franz/sud/components/engine/GameManager.java:
```diff
@@ -258,35 +254,30 @@ private void showCharacter() {
+     * Checks the item variable of the room. The block exits if the item is not in the game Inventory.
+     * Also exits if the item already exist in the hero's inventory
+     */
+    private void addRoomItem() {
```
Is the javadoc comment still correct?

zmenriquez commented on src/com/franz/sud/components/engine/GameInitializer.java:
```diff
@@ -213,16 +212,16 @@ public void setupUnits() {
         // == Construction of units ==
 
         //Hero of the game
         SkilledUnit alucard = new SkilledUnit.Builder()
```
I think you can just assign it directly to the `hero` variable.

franz-delatorre replied:
Have set the heal value in the constructor. It is already assigned during the construction of the unit. 
`SkilledUnit warlock = new SkilledUnit.Builder()
                .damage(25)
                .health(200)
                .name("Warlock")
                .evasion(10)
                .criticalChance(10)
                .lifesteal(25)
                .setSkill(heal)
                .build();`

zmenriquez replied:
Hmm, I meant instead of doing this:
```
SkilledUnit alucard = new SkilledUnit.Builder()...build();
hero = alucard;
```
You can just do this:
```
hero = new SkilledUnit.Builder()...build();
```
Since there's no other purpose for the alucard variable.

zmenriquez commented on src/com/franz/sud/components/engine/GameInitializer.java:
```diff
@@ -433,23 +431,22 @@ public void setupItems() {
+    private void setupDialogue() {
 
         // Sets the String array of each progress/room.
         String[] nullNarrative = new String[] {};
```
What's the purpose of `nullNarrative`?

I can't comment in the appropriate lines so I'll put it here:
* There's a typo in `foundAnItem`, `enemySlain`, etc.
* ```
   String[] enemyFound = new String[] {
                "Watch out! There's a banshee waiting for you at the corner",
   ```
   Is there a chance it will show this even if it's a different enemy?

zmenriquez commented on src/com/franz/sud/components/engine/GameManager.java:
```diff
@@ -77,10 +79,9 @@ private void getUserAction() {
-                map.showMap();
                 move();
                 break;
             case "c":
```
I can't comment on Line 90 so I'm putting it here. Please move the `default` case at the end of the switch.

Put the code block in case `q` in a different method.

zmenriquez commented on src/com/franz/sud/components/engine/GameManager.java:
```diff
@@ -109,6 +110,7 @@ private void getUserAction() {
      * it is not listed in the open rooms of the current map used.
      */
     private void move() {
```
Instead of recursion, convert this to a loop and remove calls of `move()` in `moveTo()`.

Commit:
SUD: asd

zmenriquez commented on src/com/franz/sud/components/engine/GameManager.java:
```diff
@@ -306,32 +297,22 @@ private void checkForEnemy() {
-        if (index > 0 ) {
-            n.setNarrated(true);
-        }
+        if (index == MAX_NARRATIVE ) n.setNarrated(true);
```
Exhibit A:
```
if (n.isNarrated()) {
            return;
}
```
Exhibit B:
```
if (index == MAX_NARRATIVE ) n.setNarrated(true);
```
Be consistent on how you format your code. 

But honestly, I like the first one better.

Commit:
SUD: made changes based on the new game architecture

franz-delatorre commented:
Have addressed the issues. 
In addition, you might notice some changes in the architecture. These changes are based on what me and Cris discussed recently of changing the game Engine to cater cartridges. Will add in the details on the PR description. Thanks!

Commit:
SUD:
